File: 2023/day07.cpp
// https://adventofcode.com/2023/day/7

#include "../common/common.h"
#include "../common/rust.h"

#include <algorithm>
#include <array>
#include <compare>
#include <print>
#include <ranges>

using Input = Vec<std::pair<String, i32>>;

auto parse(String const& filename) -> Input {
  using Pair = Input::value_type;
  return aoc::views::read_lines(filename) |
         stdv::transform([](str line) {
           auto [hand_str, bid_str] = aoc::split_once(line, ' ');
           return Pair{String{hand_str}, aoc::to_number<i32>(bid_str)};
         }) |
         aoc::collect_vec<Pair>();
}

constexpr let joker_card_number = 1;
constexpr let ace_card_number = 14;

template <bool joker>
fn card_to_number(char c) -> i32 {
  if (aoc::is_number(c)) {
    let number = static_cast<i32>(c - '0');
    AOC_ASSERT((number > joker_card_number) && (number < 10),
               "Invalid card number");
    return number;
  }
  switch (c) {
    case 'T':
      return 10;
    case 'J':
      if constexpr (joker) {
        return joker_card_number;
      } else {
        return 11;
      }
    case 'Q':
      return 12;
    case 'K':
      return 13;
    case 'A':
      return ace_card_number;
    default:
      throw std::runtime_error("Invalid card number");
  }
}

static_assert(7 == card_to_number<false>('7'));
static_assert(13 == card_to_number<false>('K'));
static_assert(11 == card_to_number<false>('J'));
static_assert(1 == card_to_number<true>('J'));

enum class type_t {
  high_card,
  one_pair,
  two_pair,
  three_kind,
  full_house,
  four_kind,
  five_kind,
};

template <bool joker>
struct hand_type {
  using cards_t = std::array<i32, 5>;

  cards_t cards;

  constexpr hand_type(str cards_str) {
    stdr::transform(cards_str, cards.begin(), &card_to_number<joker>);
  }

  constexpr hand_type(const cards_t& cards_) : cards{cards_} {}

  fn type() const -> type_t {
    std::array<i32, ace_card_number> card_map{};
    i32 num_jokers = 0;
    for (auto c : cards) {
      if (c == joker_card_number) {
        if constexpr (joker) {
          ++num_jokers;
        } else {
          // This shouldn't actually be triggered,
          // we're just preventing code from being generated unnecessarily
        }
      } else {
        // Minus 2 to exclude 0 and 1
        ++card_map[c - 2];
      }
    }
    stdr::sort(card_map, std::greater<>{});
    if constexpr (joker) {
      card_map[0] += num_jokers;
    }
    switch (card_map[0]) {
      case 5:
        return type_t::five_kind;
      case 4:
        return type_t::four_kind;
      case 3:
        if (card_map[1] == 2) {
          return type_t::full_house;
        } else {
          return type_t::three_kind;
        }
      case 2:
        if (card_map[1] == 2) {
          return type_t::two_pair;
        } else {
          return type_t::one_pair;
        }
      default:
        return type_t::high_card;
    }
  }

  fn operator<=>(const hand_type& other) const->std::strong_ordering {
    auto type_comp = (type() <=> other.type());
    if (type_comp != std::strong_ordering::equal) {
      return type_comp;
    }
    return std::strong_order(cards, other.cards);
  }

  constexpr bool operator==(const hand_type& other) const = default;
};

static_assert(hand_type<false>{"AAAAA"}.type() == type_t::five_kind);
static_assert(hand_type<false>{"AA8AA"}.type() == type_t::four_kind);
static_assert(hand_type<false>{"23332"}.type() == type_t::full_house);
static_assert(hand_type<false>{"TTT98"}.type() == type_t::three_kind);
static_assert(hand_type<false>{"23432"}.type() == type_t::two_pair);
static_assert(hand_type<false>{"A23A4"}.type() == type_t::one_pair);
static_assert(hand_type<false>{"23456"}.type() == type_t::high_card);

static_assert(hand_type<false>{"T55J5"}.type() == type_t::three_kind);
static_assert(hand_type<true>{"T55J5"}.type() == type_t::four_kind);

static_assert(hand_type<false>{"23332"} > hand_type<false>{"TTT98"});
static_assert(hand_type<false>{"33332"} > hand_type<false>{"2AAAA"});
static_assert(hand_type<false>{"77888"} > hand_type<false>{"77788"});

static_assert(hand_type<false>{"KTJJT"} < hand_type<false>{"KK677"});
static_assert(hand_type<true>{"KTJJT"} > hand_type<true>{"KK677"});

template <bool joker>
using bid_type = std::pair<hand_type<joker>, i32>;

template <bool joker>
fn total_winnings(Vec<bid_type<joker>> const& bids) -> i32 {
  let sorted_bids = aoc::ranges::sorted(bids);
  return aoc::ranges::accumulate(
      sorted_bids | stdv::enumerate | stdv::transform([](let& elem) {
        let & [ i, bid ] = elem;
        return static_cast<i32>(i + 1) * bid.second;
      }),
      0);
}

static_assert((765 * 1 + 12 * 2) ==
              total_winnings<false>({bid_type<false>{"KTJJT", 765},
                                     bid_type<false>{"QQQJA", 12}}));
static_assert((765 * 2 + 12 * 1) ==
              total_winnings<true>({bid_type<true>{"KTJJT", 765},
                                    bid_type<true>{"QQQJA", 12}}));

template <bool joker>
fn solve_case(Input const& raw) -> i32 {
  return total_winnings<joker>( //
      raw |
      stdv::transform([](let& elem) {
        let & [ hand_str, bid ] = elem;
        return bid_type<joker>(hand_str, bid);
      }) |
      aoc::collect_vec<bid_type<joker>>());
}

int main() {
  std::println("Part 1");
  let example = parse("day07.example");
  AOC_EXPECT_RESULT(6440, (solve_case<false>(example)));
  let input = parse("day07.input");
  AOC_EXPECT_RESULT(249638405, (solve_case<false>(input)));

  std::println("Part 2");
  AOC_EXPECT_RESULT(5905, (solve_case<true>(example)));
  AOC_EXPECT_RESULT(249776650, (solve_case<true>(input)));

  AOC_RETURN_CHECK_RESULT();
}

File: rust/2023/src/day07.rs
type Input = Vec<(String, i32)>;

fn parse(filename: &str) -> Input {
    aoc::file::read_lines(filename)
        .into_iter()
        .map(|line| {
            let (hand_str, bid_str) = line.split_once(' ').unwrap();
            (hand_str.to_string(), bid_str.parse().unwrap())
        })
        .collect()
}

const JOKER_CARD_NUMBER: i32 = 1;
const ACE_CARD_NUMBER: usize = 14;

fn card_to_number<const JOKER: bool>(c: char) -> i32 {
    if c.is_ascii_digit() {
        let number = c as i32 - '0' as i32;
        debug_assert!(
            number > JOKER_CARD_NUMBER && number < 10,
            "Invalid card number"
        );
        return number;
    }
    match c {
        'T' => 10,
        'J' => {
            if JOKER {
                JOKER_CARD_NUMBER
            } else {
                11
            }
        }
        'Q' => 12,
        'K' => 13,
        'A' => ACE_CARD_NUMBER as i32,
        _ => panic!("Invalid card number"),
    }
}

#[derive(PartialEq, Eq, PartialOrd, Ord, Clone, Copy)]
enum HandKind {
    HighCard,
    OnePair,
    TwoPair,
    ThreeKind,
    FullHouse,
    FourKind,
    FiveKind,
}

#[derive(Clone, Copy, PartialEq, Eq)]
struct Hand<const JOKER: bool> {
    cards: [i32; 5],
}

impl<const JOKER: bool> Hand<JOKER> {
    fn new(cards_str: &str) -> Self {
        let mut cards = [0i32; 5];
        for (i, c) in cards_str.chars().enumerate() {
            cards[i] = card_to_number::<JOKER>(c);
        }
        Hand { cards }
    }

    fn kind(&self) -> HandKind {
        let mut card_map = [0i32; ACE_CARD_NUMBER];
        let mut num_jokers = 0i32;
        for &c in &self.cards {
            if c == JOKER_CARD_NUMBER {
                if JOKER {
                    num_jokers += 1;
                }
            } else {
                // Minus 2 to exclude 0 and 1
                card_map[(c - 2) as usize] += 1;
            }
        }
        card_map.sort_unstable_by(|a, b| b.cmp(a));
        if JOKER {
            card_map[0] += num_jokers;
        }
        match card_map[0] {
            5 => HandKind::FiveKind,
            4 => HandKind::FourKind,
            3 => {
                if card_map[1] == 2 {
                    HandKind::FullHouse
                } else {
                    HandKind::ThreeKind
                }
            }
            2 => {
                if card_map[1] == 2 {
                    HandKind::TwoPair
                } else {
                    HandKind::OnePair
                }
            }
            _ => HandKind::HighCard,
        }
    }
}

impl<const JOKER: bool> Ord for Hand<JOKER> {
    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
        self.kind()
            .cmp(&other.kind())
            .then_with(|| self.cards.cmp(&other.cards))
    }
}

impl<const JOKER: bool> PartialOrd for Hand<JOKER> {
    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
        Some(self.cmp(other))
    }
}

fn total_winnings<const JOKER: bool>(mut bids: Vec<(Hand<JOKER>, i32)>) -> i32 {
    bids.sort_unstable();
    bids.iter()
        .enumerate()
        .map(|(i, (_, bid))| (i as i32 + 1) * *bid)
        .sum()
}

fn solve_case<const JOKER: bool>(input: &Input) -> i32 {
    let bids = input
        .iter()
        .map(|(hand_str, bid)| (Hand::<JOKER>::new(hand_str), *bid))
        .collect();
    total_winnings(bids)
}

fn main() {
    println!("Part 1");
    let example = parse("day07.example");
    aoc::expect_result!(6440, solve_case::<false>(&example));
    let input = parse("day07.input");
    aoc::expect_result!(249638405, solve_case::<false>(&input));

    println!("Part 2");
    aoc::expect_result!(5905, solve_case::<true>(&example));
    aoc::expect_result!(249776650, solve_case::<true>(&input));
}
